src/Servers/Kestrel/Core/src/Internal/Infrastructure/StringUtilities.cs

```diff
@@ -115,6 +115,93 @@ public static unsafe bool TryGetAsciiString(byte* input, char* output, int count
+        public static unsafe void GetLatin1String(byte* input, char* output, int count)
+        {
+            // Calculate end position
+            var end = input + count;
+
+            do
+            {
+                // If Vector not-accelerated or remaining less than vector size
+                if (!Vector.IsHardwareAccelerated || input > end - Vector<sbyte>.Count)
+                {
+                    if (IntPtr.Size == 8) // Use Intrinsic switch for branch elimination
+                    {
+                        // 64-bit: Loop longs by default
+                        while (input <= end - sizeof(long))
+                        {
+                            output[0] = (char)input[0];
+                            output[1] = (char)input[1];
+                            output[2] = (char)input[2];
+                            output[3] = (char)input[3];
+                            output[4] = (char)input[4];
+                            output[5] = (char)input[5];
+                            output[6] = (char)input[6];
+                            output[7] = (char)input[7];
+
+                            input += sizeof(long);
+                            output += sizeof(long);
+                        }
+                        if (input <= end - sizeof(int))
+                        {
+                            output[0] = (char)input[0];
+                            output[1] = (char)input[1];
+                            output[2] = (char)input[2];
+                            output[3] = (char)input[3];
+
+                            input += sizeof(int);
+                            output += sizeof(int);
+                        }
+                    }
+                    else
+                    {
+                        // 32-bit: Loop ints by default
+                        while (input <= end - sizeof(int))
+                        {
+                            output[0] = (char)input[0];
+                            output[1] = (char)input[1];
+                            output[2] = (char)input[2];
+                            output[3] = (char)input[3];
+
+                            input += sizeof(int);
+                            output += sizeof(int);
+                        }
+                    }
+                    if (input <= end - sizeof(short))
+                    {
+                        output[0] = (char)input[0];
+                        output[1] = (char)input[1];
+
+                        input += sizeof(short);
+                        output += sizeof(short);
+                    }
+                    if (input < end)
+                    {
+                        output[0] = (char)input[0];
+                    }
+
+                    return;
+                }
+
+                // do/while as entry condition already checked
+                do
+                {
+                    var vector = Unsafe.AsRef<Vector<sbyte>>(input);
+                    Vector.Widen(
```

Comment:
If this method is hot, you could add explicit paths for AVX2, SSE2, because
* Vector operates on the widest op-set available, i.e. AVX, so some iterations need to be done sequentially instead of parallel as SSE would allow
* Vector-methods may result in somewhat not ideal codegen.

For reference see https://github.com/dotnet/aspnetcore/blob/80a77ff5c23647fa52782d15d0abb92cd2afbbb1/src/Shared/ServerInfrastructure/StringUtilities.cs#L19 from https://github.com/dotnet/aspnetcore/pull/17556

Except `CheckBytesInAsciiRange` both methods are similar. So they could be combined to one, and with some tricks let the JIT generate the proper code for both flavors.
I could do this as part of my PR, if this one gets merged first (also add the BMI-path, comment above).

Reply:
We don't expect this method to be hot. This is mostly here so Bing can accept non-UTF8/ASCII encoded headers from legacy clients. This is opt-in via an undocumented IConfiguration flag. Even when opted-in, this method will only be hit to decode non-ASCII request headers.

I'll look into merging #17556 to master soon. I'll convert GetLatin1String to use your TryGetAsciiString optimizations when I do.

Reply:
That is if we don't just end up using Encoding.Latin1 in .NET 5 which we very well might.